apis/python/src/tiledbsoma/io/ingest.py

```diff
@@ -1351,19 +1355,57 @@ def to_anndata(
             shape = measurement.obsm[key].shape
             if len(shape) != 2:
                 raise ValueError(f"expected shape == 2; got {shape}")
-            matrix = measurement.obsm[key].read((slice(None),) * len(shape)).to_numpy()
-            # The spelling ``sp.csr_array`` is more idiomatic but doesn't exist until Python 3.8
-            obsm[key] = sp.csr_matrix(matrix)
+            if isinstance(measurement.obsm[key], DenseNDArray):
+                matrix = (
+                    measurement.obsm[key].read((slice(None),) * len(shape)).to_numpy()
+                )
+                # The spelling ``sp.csr_array`` is more idiomatic but doesn't exist until Python 3.8
+                obsm[key] = sp.csr_matrix(matrix)
```

Comment:
The `obsm` and `varm` matrices in AnnData should be dense, not sparse. While AnnData allows sparse matrices in `obsm` and `varm`, in practice these are _never_ sparse. The most efficient storage form for these matrices is as a dense array

```python
>>> import numpy
>>> import scipy.sparse as sp
>>> from pympler import asizeof
>>>
>>> nobs = 500000
>>> npcs = 50
>>>
>>> arr = numpy.array((1.1,) * (nobs * npcs))
>>> arr.shape = (nobs, npcs)
>>>
>>> mat = sp.csr_matrix(arr)
>>>
>>> asizeof.asizeof(arr)
200000144
>>> asizeof.asizeof(mat)
302001400
```

If you were to store as sparse, then you should use `csc` instead of `csr` as access will (nearly) always be along the reduced dimension (eg. select PCs 1 and 2 for all cells), not `obs` or `var` (select all PCs for cells 3 and 4). But they should be stored as dense

Reply:
@mojaveazure fixed -- thanks! :)